py/runtime.c

```diff
@@ -1409,3 +1409,76 @@ NORETURN void mp_raise_TypeError(const char *msg) {
 NORETURN void mp_not_implemented(const char *msg) {
     mp_raise_msg(&mp_type_NotImplementedError, msg);
 }
+
+#if MICROPY_PY_SOFTIRQ
+
+#define MAX_STACK_DEPTH (5)
+
+typedef struct _stack_elem_t {
+    mp_obj_t irqfunc;
+    mp_obj_t arg1;
+    mp_obj_t arg2;
+} stack_elem_t;
+
+volatile bool softirq_enabled = false;
+volatile short sp = -1;
+stack_elem_t stack[MAX_STACK_DEPTH];
+
+void mp_enable_softirq(void) {
+    softirq_enabled = true;
+}
+
+void mp_disable_softirq(void) {
+    mp_uint_t irqstate = disable_irq();
+    softirq_enabled = false;
+    sp = -1;
+    MP_STATE_VM(mp_pending_ex_flags) &= ~PENDING_EX_SOFT_INT;
+    enable_irq(irqstate);
+}
+
+void mp_exec_softirq(void) {
+    stack_elem_t softirq;
+    mp_uint_t irqstate;
+    static bool busy = false;
+
+    if (sp >= 0 && busy == false && __get_IPSR() == 0) {
+        irqstate = disable_irq();
+        softirq = stack[sp--];
+        busy = true;
+        enable_irq(irqstate);
+        if (softirq.arg1 == MP_OBJ_NULL) {
+            mp_call_function_0(softirq.irqfunc);
+        } else if (softirq.arg2 == MP_OBJ_NULL) {
+            mp_call_function_1(softirq.irqfunc, softirq.arg1);
+        } else {
+            mp_call_function_2(softirq.irqfunc, softirq.arg1, softirq.arg2);
+        }
+        irqstate = disable_irq();
+        busy = false;
+        if (sp == -1) {
+            MP_STATE_VM(mp_pending_ex_flags) &= ~PENDING_EX_SOFT_INT;
+        }
+        enable_irq(irqstate);
+    }
+}
+
+mp_obj_t mp_add_softint(mp_obj_t function, mp_obj_t arg1, mp_obj_t arg2) {
+
+    if (sp < MAX_STACK_DEPTH - 1 && softirq_enabled) {
```

Comment:
@dpgeorge : I guess you mean the idea that SoftIRQs should follow HardIRQs in behavior closely. I'm not sure about the exact implementation though. From common sense, HardIRQ handler indeed can just set a single-bit flag in some variable for SoftIRQ, just the same as CPU sets that bit in some register for HardIRQ to happen. But our API and latency requirement may may it not that simple. For example, we attach handlers not to hardware IRQ numbers, but to specific objects, right? For example, to a specific Pin, which may share IRQ eith other pins. And with SoftIRQ processing delays, the same HardIRQ may happen again, but for another object, while first wasn't yet served. So, having a proper queue, with callback params as done by @HenrikSolver makes sense. (I don't even ask why there're slots for up to 2 args, it also seems to be optimized, up to support a handler which is a method of object and accepts a callback arg (though, that worth to be checked ;-) )) .